proposed/2022/Package-Source-Mapping-VS-Options.md

```diff
@@ -0,0 +1,133 @@
+# Package Source Mapping in Visual Studio Options Dialog
+
+* Start Date: 06-10-2022
+* Authors: Ella McNally
+* Issue: https://github.com/NuGet/Home/issues/11363
+
+## Summary
+
+Currently there is no support for package source mapping in the Visual Studio options dialog. I will be creating a feature where users can pin default sources and choose a specific source(s) for certain package source mappings from sources already provided in package sources page. This feature is a portion of what was discussed in a previous spec. https://github.com/NuGet/Home/blob/dev/implemented/2021/PackageSourceMapping.md
+
+## Motivation 
+
+Adding support for package source mapping in the VS options dialog will allow user to pin sources and add/remove package source mappings with more ease.
+
+## Explanation
+
+### Functional Explanation
+
+#### Package Source Mappings Page 
+
+When the user opens the package source mapping page in the options dialog, they will see a list of all the package source mappings they have already configured. This page will have three buttons: _Add_ , _Remove_, and _Clear_.
+
+* Add: Brings user to a pop-up where they can add package source mappings to their project.
+
+* Remove: Lets user remove selected package source mapping from project.
+
+* Clear: Lets user clear all package source mappings from current project.
+
+The popup when the user clicks the add button will have a textbox where a user can type in a package ID. All sources that are defined in package sources page of VS options dialog will appear in a `ComboBox` below. The user checks a `CheckBox` for the sources they would like to use for that package ID. 
+
+The add window has two buttons: _Add_ and _Close_.
+
+* Add: Add button adds package source mapping to project only using selected sources, and the pop-up will disappear. 
+
+* Close: Will close pop-up without adding any package source mappings.
+
+#### Package Sources Page
+
+I will add a feature that allows a user to pin default sources for all package source mappings that are not specified on the package source mapping page. There will be a pin button, that when clicked, creates a default mapping for all package IDs to the pinned source. A pin icon will appear next to all pinned sources.

```

Comment:
AFAIK, to pin a package source for all NuGet operations there should be a source mapping configured with pattern `*`. Did I miss something here?

Reply:
Kartheek, is the example under Technical Explanation what you're looking for? `pattern=*` is described there. https://github.com/NuGet/Home/pull/11865/files/a43a3e807fa606013bf97413088c8407724498f0#diff-f9e22802f1ca486df02ad7e0668bf53e5c4c087c6c5adad061653045811fdccaR86

Reply:
I'm not 100% sure the meaning of the `pin` is something easily understood. 
I understand this is partially a comment on the original mock-ups but I believe it's something up for discussion. 

I'm not sure I have another proposal for it, beyond maybe just not having it.

Reply:
@nkolev92  That's a fair point. I feel like a tooltip for the pin icon would be helpful.

Reply:
I would propose not even having it in the first place. 
I feel like the concept isn't worth special casing:

Say you have the following: 

nuget.org - *

contoso - special.id
contoso - Microsoft.*
contoso - nuget.*
contoso - contoso.*

It's super likely that most packages are getting restored from contoso, but nuget.org will have the pin, and contoso wouldn't. 

Having something telling us about nuget.org having the fallback version can even be misleading.

Reply:
I wonder if we should have something saying a mapping exists for a certian package source? That might be more helpful.

Reply:
I think of the pin as a graphical representation that this exists:
`nuget.org - *`

Are you thinking that's not worth representing in a UI, or are can you elaborate if you're saying more than that?

Reply:
> It's super likely that most packages are getting restored from contoso, but nuget.org will have the pin, and contoso wouldn't.

If Ella's proposal for the tooltip on the pin emphasizes the meaning of Default, would that alleviate the concern?
I think we're going this direction because once we modify PMUI to support Pinning, the concepts will amplify the point of a pin icon.

Reply:
> Are you thinking that's not worth representing in a UI, or are can you elaborate if you're saying more than that?

In my opinion, the package sources page should not have any representation about which source is default. This is assuming that the package source mapping config is shown in a different page.  The package source mapping page having an icon or something to show it's the default seems reasonable.

Going off on a tangent, [the feature spec for package source mapping](https://github.com/NuGet/Home/blob/dev/implemented/2021/PackageSourceMapping.md) uses the word "pin" in the context that a package that has a mapping is now pinned to a source. Reusing the same word, pin, to mean something different (package source used when no other package prefix matches) risks confusion and bad SEO, assuming we as a team talk to customers using this terminology, and customers also use it. It would be ideal if we could agree on standard terminology and use it consistently to improve searchability and reduce confusion.

Reply:
Another option for the package sources dialog:

Add an indicator for each source that has a mapping defined. 

That indicator could have a tooltip that could call out the default/fallback concept if we wanted to.
It can also just say `see the package source mapping page for mapping information`.

Reply:
Reading other comments in this PR, it reminded me that a package source with prefix `*` is a fallback source, not really a default source. The `*` prefix will be used if no other prefix matches (longest prefix wins). To me, that makes the package source feel less important, so maybe it doesn't warrant to be pointed out more clearly than other package sources?

Reply:
> , [the feature spec for package source mapping](https://github.com/NuGet/Home/blob/dev/implemented/2021/PackageSourceMapping.md) uses the word "pin" in the context that a package that has a mapping is now pinned to a source. Reusing the same word, pin, to mean something different (package source used when no other package prefix matches) risks confusion and bad SEO,

If you scroll down to the Mockups section of the file you linked, you'll see the pin icon and the terminology of pin used there. We didn't deviate from the feature spec here in the UI spec. Whether we want to change the feature spec's direction is debatable. 

For now, we'll save changing the sources page to last. I think that's not blocking anything on this spec, and it's more easily changeable as we iterate.

Reply:
> Reading other comments in this PR, it reminded me that a package source with prefix `*` is a fallback source, not really a default source. The `*` prefix will be used if no other prefix matches (longest prefix wins). To me, that makes the package source feel less important, so maybe it doesn't warrant to be pointed out more clearly than other package sources?

100% agree with this. Better summary of my concerns.

Reply:
the merged spec doesn't really call out the fact that we're not going to add the pin thumbthack in V1. 
It's not a big deal, given that we have a tracking issue for it, but we can always just link it.